alchemist/alchemist-incarnation-biochemistry/src/main/java/it/unibo/alchemist/model/implementations/environments/BioRect2DEnvironmentNoOverlap.java

```diff
@@ -0,0 +1,263 @@
+        final double module =  FastMath.sqrt(FastMath.pow((yIntersect - yo), 2) + FastMath.pow((xIntersect - xo), 2));
+        if (module == 0) { // if module == 0 the translation is 0, so return originalPos
+            return Optional.of(originalPos);
+        }
+        // compute the versor relative to requested direction of cell movement
+        final double cat2 = FastMath.sqrt((FastMath.pow(cellRange, 2) - FastMath.pow(cat, 2)));
+        final double distToSum  = originalPos.getDistanceTo(intersection) - cat2;
+        final Position versor = new Continuous2DEuclidean((xIntersect - xo) / module, (yIntersect - yo) / module);
+        // computes vector representing the practicable movement
+        final Position vectorToSum = new Continuous2DEuclidean(distToSum * (versor.getCoordinate(0)), distToSum * (versor.getCoordinate(1)));
+        // returns the right position of the cell
+        final Position result = originalPos.sum(vectorToSum);
+        return Optional.of(result);
+    }
+
+    @Override
+    protected void nodeAdded(final Node<Double> node, final Position position, final Neighborhood<Double> neighborhood) {
+        super.nodeAdded(node, position, neighborhood);
+        if (node instanceof CellWithCircularArea) {
+            final CellWithCircularArea cell = (CellWithCircularArea) node;
+            if (cell.getDiameter() > getMaxDiameterAmongCells()) {
+                biggestCell = Optional.of(cell); 
+            }
+        }
+        if (node instanceof CircularDeformableCell) {
+            final CircularDeformableCell cell = (CircularDeformableCell) node;
+            if (cell.getMaxDiameter() > getMaxDiameterAmongDeformableCells()) {
+                biggestDeformableCell = Optional.of(cell); 
+            }
+        }
+    }
+
+    @Override
+    protected void nodeRemoved(final Node<Double> node, final Neighborhood<Double> neighborhood) {
+        if (biggestCell.isPresent() && biggestCell.get().equals(node)) {
+            biggestCell = getNodes().stream()
+                    .parallel()
+                    .flatMap(n -> n instanceof CellWithCircularArea ? Stream.of((CellWithCircularArea) n) : Stream.empty())
+                    .max((c1, c2) -> Double.compare(c1.getDiameter(), c2.getDiameter()))
+                    .map(Optional::of)
+                    .orElse(Optional.absent());
+        }
+        if (biggestDeformableCell.isPresent() && biggestDeformableCell.get().equals(node)) {
+            biggestDeformableCell = getNodes().stream()
+                    .parallel()
+                    .flatMap(n -> n instanceof CircularDeformableCell ? Stream.of((CircularDeformableCell) n) : Stream.empty())
+                    .max((c1, c2) -> Double.compare(c1.getMaxDiameter(), c2.getMaxDiameter()))
+                    .map(Optional::of)
+                    .orElse(Optional.absent());
+        }
+    }
+
+    private double getMaxDiameterAmongCells() {
```

Comment:
Refactor. E.g., use a new private method with a parameter, that applies transformation and or, then call it in `getMaxDiameterAmongCells` and `getMaxDiameterAmongDeformableCells`